Cancel an Activity from a Workflow (#1938)
temporalio/documentation: docs-src/app-dev-context/cancel-activity.md

---
id: cancel-activity
title: Cancel an Activity from a Workflow

> description: If an Activity is supposed to react to Cancellation.

Reviewer: ```suggestion
description: An Activity can be canceled from within a Workflow if the Activity sends Heartbeats.
```

sidebar_label: Cancel an Activity
tags:
  - guide-context
---

> Cancelling an Activity from within a Workflow requires that the Activity Execution sends Heartbeats and sets a Heartbeat Timeout. If the Heartbeat is not invoked, the Activity cannot receive a cancellation request. Users should set Heartbeat and set a [Heartbeat Timeout](/concepts/what-is-a-heartbeat-timeout) when executing any non-immediate Activity to ensure the server knows it is still working.

Reviewer: ```suggestion
Canceling an Activity from within a Workflow requires that the Activity Execution sends Heartbeats and sets a Heartbeat Timeout.
If the Heartbeat is not invoked, the Activity cannot receive a cancellation request.
When any non-immediate Activity is executed, the Activity Execution should send Heartbeats and set a [Heartbeat Timeout](/concepts/what-is-a-heartbeat-timeout) to ensure that the server knows it is still working.
```

> When an Activity is cancelled, an error is raised in the Activity at the next available opportunity. Temporal recommends catching the error and perform clean-up logic to ensure that the task is properly completed.

Reviewer 2: ```suggestion
When an Activity is cancelled, an error is raised in the Activity at the next available opportunity. If cleanup logic needs to be performed, it can be done in a finally or inside a caught cancel error, but in order for the activity to appear cancelled the exception needs to be re-raised.
```

We definitely never want recommending catching the error unless they absolutely have to.

> When an Activity is cancelled, an error is raised in the Activity at the next available opportunity. If cleanup logic needs to be performed, it can be done in a `finally` clause or inside a caught cancel error, but in order for the Activity to appear cancelled the exception needs to be re-raised.

Reviewer: ```suggestion
When an Activity is canceled, an error is raised in the Activity at the next available opportunity.
If cleanup logic needs to be performed, it can be done in a `finally` clause or inside a caught cancel error.
However, for the Activity to appear canceled the exception needs to be re-raised.
```

:::note

> Unlike regular Activities, Core-based [Local Activities](/concepts/what-is-a-local-activity) do not require Heartbeats to be cancelled. This is because they are handled locally and all the information needed to handle the cancellation logic is available in the same Worker process. As a result, Local Activities can be cancelled without the need for Heartbeats.

Reviewer 2: ```suggestion
Unlike regular Activities, [Local Activities](/concepts/what-is-a-local-activity) do not require Heartbeats to be cancelled. This is because they are handled locally and all the information needed to handle the cancellation logic is available in the same Worker process. As a result, Local Activities can be cancelled without the need for Heartbeats.
```

> Unlike regular Activities, [Local Activities](/concepts/what-is-a-local-activity) do not require Heartbeats to be cancelled. This is because they are handled locally and all the information needed to handle the cancellation logic is available in the same Worker process. As a result, Local Activities can be cancelled without the need for Heartbeats.

Reviewer: ```suggestion
Unlike regular Activities, [Local Activities](/concepts/what-is-a-local-activity) can be canceled if they don't send Heartbeats.
Local Activities are handled locally, and all the information needed to handle the cancellation logic is available in the same Worker process.
```
